### ISSUE TITLE
systemd private library libsystemd-shared-252.so not found by 'ldd'

### ISSUE THREAD

**Issue by LiamFry** [rear/rear#3021]
Backup (`rear -v mkbackup`) reports that a necessary file is missing. I don't understand ...

* ReaR version: Relax-and-Recover 2.7 / Git
* OS version: Debian GNU/Linux 12 (bookworm)
* ReaR configuration files: (attached)
* System architecture: x86_64
* Storage layout: (attached)
* Description of the issue:
I am not sure if this is a `rear` issue yet I cannot find any useful information elsewhere. (I've tried!) During backup, I get the following message ...
```
Testing that the recovery system in /var/tmp/rear.PvzABq2iZtXbVTm/rootfs contains a usable system
/usr/lib/x86_64-linux-gnu/systemd/libsystemd-core-252.so requires additional libraries
        libsystemd-shared-252.so => not found
ReaR recovery system in '/var/tmp/rear.PvzABq2iZtXbVTm/rootfs' needs additional libraries, check /var/log/rear/rear-clara.log for details
```
`dpkg --get-selections | grep libsystemd-shared` shows that it's installed.
`dpkg -L libsystemd-shared` gives me the following ...
```
/.
/usr
/usr/lib
/usr/lib/x86_64-linux-gnu
/usr/lib/x86_64-linux-gnu/systemd
/usr/lib/x86_64-linux-gnu/systemd/libsystemd-core-252.so
/usr/lib/x86_64-linux-gnu/systemd/libsystemd-shared-252.so
/usr/share
/usr/share/doc
/usr/share/doc/libsystemd-shared
/usr/share/doc/libsystemd-shared/NEWS.Debian.gz
/usr/share/doc/libsystemd-shared/changelog.Debian.gz
/usr/share/doc/libsystemd-shared/copyright
/usr/share/lintian
/usr/share/lintian/overrides
/usr/share/lintian/overrides/libsystemd-shared
```
All that said, a quick `find / -name libsystemd-shared-252.so -ls` reports this:
```
22458549 3252 -rw-r--r-- 1 root root 3327624 Feb 28 06:15 /usr/lib/x86_64-linux-gnu/systemd/libsystemd-shared-252.so
```

[local.conf.txt](https://github.com/rear/rear/files/11878771/local.conf.txt)
[storage-layout.txt](https://github.com/rear/rear/files/11878819/storage-layout.txt)

**Comment by jsmeix** [comment]
I am not a Debian user
so I cannot reproduce Debian specific issues.

In general when things are missing in the ReaR recovery system
you could manually add what is needed via something like
```
LIBS+=( /usr/lib/x86_64-linux-gnu/systemd/libsystemd-shared-252.so )
```
or
```
LIBS+=( /usr/lib/x86_64-linux-gnu/systemd/*.so* )
```
in your etc/rear/local.conf file.

See the descriptions about LIBS and COPY_AS_IS
in usr/share/rear/conf/default.conf

**Comment by pcahyna** [comment]
I see a similar message on Fedora 37:
```
Testing that the recovery system in /tmp/rear.OBGH5kwM9RvJv5S/rootfs contains a usable system
There are binaries or libraries in the ReaR recovery system that need additional libraries
/usr/lib/systemd/libsystemd-core-251.14-2.fc37.so requires additional libraries
libsystemd-shared-251.14-2.fc37.so => not found
```
Weird.

**Comment by LiamFry** [comment]
A quick run of `ldd /usr/lib/x86_64-linux-gnu/systemd/libsystemd-core-252.so` did report that `libsystemd-shared-252.so` was missing. This is damned odd.
I took my question to the Debian forums. I'll post here what I discover.

**Comment by pcahyna** [comment]
Similar on Fedora 37:
```
# ldd /usr/lib/systemd/libsystemd-core-251.14-2.fc37.so 
        linux-vdso.so.1 (0x00007ffefb5d6000)
        libsystemd-shared-251.14-2.fc37.so => not found
```
libsystemd-shared-251.14-2.fc37.so is actually in /usr/lib/systemd .

**Comment by jsmeix** [comment]
I guess that the RequiredSharedObjects function in ReaR
does not get called for libsystemd-core-252.so
so libsystemd-shared-252.so is not automatically
included into the ReaR recovery system.

RequiredSharedObjects is called in
build/GNU/Linux/390_copy_binaries_libraries.sh
for all existing programs in PROGS and REQUIRED_PROGS
and for all libraries in LIBS via
```
local all_libs=( "${LIBS[@]}" $( RequiredSharedObjects "${all_binaries[@]}" "${LIBS[@]}" ) )
```
and RequiredSharedObjects is called in
build/GNU/Linux/100_copy_as_is.sh
for all executable files in COPY_AS_IS.

So when libsystemd-core-252.so is not specified in LIBS
but gets included into the recovery system via COPY_AS_IS
and when libsystemd-core-252.so is not an executable file
then RequiredSharedObjects is not called for libsystemd-core-252.so
so libsystemd-core-252.so is not automatically included
and when libsystemd-shared-252.so is not explicitly
specified to be included into the recovery system
then libsystemd-shared-252.so is missing in the recovery system.

**Comment by jsmeix** [comment]
@pcahyna 
Ah!
It seems the library configuration on the original system
could be incomplete when `ldd` does not find it.

**Comment by jsmeix** [comment]
FYI:
On my current openSUSE Leap 15.4 system
I have systemd-249.16 which contains only
/usr/lib/systemd/libsystemd-shared-249.so
and I noticed no issues with that in ReaR.

**Comment by LiamFry** [comment]
I found [this post](https://answers.launchpad.net/ubuntu/+question/703825) that offered some "solutions." Even though the question was asked for Ubuntu, I found it helpful for Debian.

TL;DR

One-off solution:
`sudo bash -c "LD_PRELOAD=/usr/lib/x86_64-linux-gnu/systemd/libsystemd-shared-252.so" rear -v mkbackup`
(untested)
... OR ...
create an file in `/etc/ld.so.conf.d`
(This is the method I chose.)

I created `/etc/ld.so.conf.d/libsystemd-core.conf`
... into which I put `/usr/lib/x86_64-linux-gnu/systemd`
I then ran `sudo ldconf`

A run of `ldd /usr/lib/x86_64-linux-gnu/systemd/libsystemd-core-252.so` now resolves `libsystemd-shared-252.so`

I rebooted for good measure and ran a straight `rear -v mkbackup` -- it ran flawlessly.

I suppose why `libsystemd-core-252.so` has this issue is a question for the distro and subsystem teams.

**Comment by ekool** [comment]
I just want to add that the above post from LiamFry fixed it for me as well.

**Comment by sid-the-sloth** [comment]
> sudo ldconf

This solution from @LiamFry  worked -- but on Debian should do `sudo ldconfig` instead.

**Comment by LiamFry** [comment]
> > sudo ldconf
> 
> This solution from @LiamFry worked -- but on Debian should do `sudo ldconfig` instead.

That's for the correction! I updated my "potential solution," above.

**Comment by zDEFz** [comment]
I need support for Fedora 38.
I tried above methods and I couldn't proceed.

```
ldd /usr/lib/systemd/libsystemd-shared-253.10-1.fc38.so
	linux-gate.so.1 (0xf7ee6000)
	libacl.so.1 => /lib/libacl.so.1 (0xf7ea8000)
	libblkid.so.1 => /lib/libblkid.so.1 (0xf7e6c000)
	libcap.so.2 => /lib/libcap.so.2 (0xf7e61000)
	libcrypt.so.2 => /lib/libcrypt.so.2 (0xf7e25000)
	libkmod.so.2 => /lib/libkmod.so.2 (0xf7e04000)
	liblz4.so.1 => /lib/liblz4.so.1 (0xf7de2000)
	libmount.so.1 => /lib/libmount.so.1 (0xf79b2000)
	libcrypto.so.3 => /lib/libcrypto.so.3 (0xf7400000)
	libpam.so.0 => /lib/libpam.so.0 (0xf7dcf000)
	libseccomp.so.2 => /lib/libseccomp.so.2 (0xf7991000)
	libselinux.so.1 => /lib/libselinux.so.1 (0xf7960000)
	liblzma.so.5 => /lib/liblzma.so.5 (0xf7928000)
	libzstd.so.1 => /lib/libzstd.so.1 (0xf7879000)
	libm.so.6 => /lib/libm.so.6 (0xf732d000)
	libgcc_s.so.1 => /lib/libgcc_s.so.1 (0xf7850000)
	libc.so.6 => /lib/libc.so.6 (0xf713f000)
	/lib/ld-linux.so.2 (0xf7ee8000)
	libattr.so.1 => /lib/libattr.so.1 (0xf7dc6000)
	libz.so.1 => /lib/libz.so.1 (0xf7836000)
	libaudit.so.1 => /lib/libaudit.so.1 (0xf7807000)
	libeconf.so.0 => /lib/libeconf.so.0 (0xf7131000)
	libpcre2-8.so.0 => /lib/libpcre2-8.so.0 (0xf7095000)
	libcap-ng.so.0 => /lib/libcap-ng.so.0 (0xf7dbb000)
```

```
Relax-and-Recover 2.7 / Git
Running rear mkrescue (PID 7919 date 2023-10-05 20:20:28)
Using log file: /var/log/rear/rear-mainrig-MS-7D51.log
Running workflow mkrescue on the normal/original system
Using UEFI Boot Loader for Linux (USING_UEFI_BOOTLOADER=1)
Using autodetected kernel '/boot/vmlinuz-6.5.5-200.fc38.x86_64' as kernel in the recovery system
Creating disk layout
Overwriting existing disk layout file /var/lib/rear/layout/disklayout.conf
Ignoring non-existing btrfs subvolume listed as mounted: /var/lib/snapd/snap
Disabling component 'disk /dev/sda' in /var/lib/rear/layout/disklayout.conf
Disabling component 'part /dev/sda' in /var/lib/rear/layout/disklayout.conf
Disabling component 'disk /dev/sdb' in /var/lib/rear/layout/disklayout.conf
Disabling component 'part /dev/sdb' in /var/lib/rear/layout/disklayout.conf
Disabling component 'disk /dev/sdc' in /var/lib/rear/layout/disklayout.conf
Disabling component 'part /dev/sdc' in /var/lib/rear/layout/disklayout.conf
Disabling component 'disk /dev/sdd' in /var/lib/rear/layout/disklayout.conf
Disabling component 'part /dev/sdd' in /var/lib/rear/layout/disklayout.conf
Disabling component 'disk /dev/sde' in /var/lib/rear/layout/disklayout.conf
Disabling component 'part /dev/sde' in /var/lib/rear/layout/disklayout.conf
Disabling component 'fs ... /mnt/external-ssd-storage' in /var/lib/rear/layout/disklayout.conf
Disabling component 'fs ... /mnt/external-sd-storage/pianolessons' in /var/lib/rear/layout/disklayout.conf
Disabling component 'btrfsmountedsubvol ... /mnt/external-ssd-storage' in /var/lib/rear/layout/disklayout.conf
Using guessed bootloader 'EFI' for 'rear recover' (found in first bytes on /dev/nvme0n1)
Verifying that the entries in /var/lib/rear/layout/disklayout.conf are correct
Created disk layout (check the results in /var/lib/rear/layout/disklayout.conf)
Creating recovery system root filesystem skeleton layout
Skipping 'lo': not bound to any physical interface.
Skipping 'wg-mullvad': not bound to any physical interface.
To log into the recovery system via ssh set up /root/.ssh/authorized_keys or specify SSH_ROOT_PASSWORD
Trying to find what to use as UEFI bootloader...
Trying to find a 'well known file' to be used as UEFI bootloader...
Using '/boot/efi/EFI/fedora/grubx64.efi' as UEFI bootloader file (non Secure Boot)
Copying logfile /var/log/rear/rear-mainrig-MS-7D51.log into initramfs as '/tmp/rear-mainrig-MS-7D51-partial-2023-10-05T20:20:32+02:00.log'
Copying files and directories
Copying binaries and libraries
Copying all kernel modules in /lib/modules/6.5.5-200.fc38.x86_64 (MODULES contains 'all_modules')
Copying all files in /lib*/firmware/
Testing that the recovery system in /var/tmp/rear.UlMhZTMIkXksvXn/rootfs contains a usable system
/usr/lib64/systemd/libsystemd-core-253.10-1.fc38.so requires additional libraries
	libsystemd-shared-253.10-1.fc38.so => not found
ReaR recovery system in '/var/tmp/rear.UlMhZTMIkXksvXn/rootfs' needs additional libraries, check /var/log/rear/rear-mainrig-MS-7D51.log for details
Creating recovery/rescue system initramfs/initrd initrd.cgz with gzip default compression
Created initrd.cgz with gzip default compression (408933330 bytes) in 16 seconds
GRUB2 modules to load: ext2 fat part_gpt part_msdos
Did not find /boot/grub2/locale files (minor issue for UEFI ISO boot)
Making ISO image
Wrote ISO image: /var/lib/rear/output/rear-mainrig-MS-7D51.iso (517M)
Exiting rear mkrescue (PID 7919) and its descendant processes ...
Running exit tasks
```

**Comment by ufreier** [comment]
Yes, the solution from @LiamFry works in the sense of building a rescue system without errors. BUT ... has anyone successfully booted the system with another device than the original? I have 2 identical servers (one as a cold stand-by) each only with one nvme-ssd and I'm not able to boot  'the other one' with the rescue system (simulating a ssd crash and its replacement). Same env like LiamFry (Debian 12, ReaR 2.7) and same settings for rescue medium like in its local.conf. I tried Debian installations with Legacy and UEFI boot, OUTPUT=USB and OUTPUT=RAWDISK - each try to boot the other one results in hanging after 'Booting rescue system' (UEFI) resp. after 'Probing EDD (edd=off ..) ok'. (Legacy) with no errors but only power reset helps. However, the server where the rescue medium was created boots both (USB and dd'ed RAWDISK). So I suspect there is something on the original device that is needed for booting the rescue medium, that would be worse in case of a real crash. I only have this problem in Debian 12, f.i. in Ubuntu 22.04 ReaR works like expected but I don't want to switch ...

**Comment by jsmeix** [comment]
@ufreier 
only as a side note FYI regarding
```
hanging after 'Probing EDD (edd=off ..) ok'
```
see
https://github.com/rear/rear/issues/2955#issuecomment-1465642166

Check your ReaR recovery system kernel command line arguments
with your Debian 12 versus with your Ubuntu 22.04
and look for "console=..." settings.

E.g. for BIOS you can get the
ReaR recovery system kernel command line arguments
by selecting in the ReaR recovery system boot menue
the topmost enty of the form "Recover HOSTNAME" and then
pressing the [Tab] key which shows the kernel command line.

**Comment by pcahyna** [comment]
@ufreier I suspect that the issue you are encountering is unrelated to the warning message about libsystemd-shared-252.so not found. Can you please open a separate report?

**Comment by ufreier** [comment]
@jsmeix : thanks a lot, USE_SERIAL_CONSOLE='no'  was the solution!

**Comment by jsmeix** [comment]
@ufreier 
you are welcome

**Comment by pcahyna** [comment]
@ufreier how old is ReaR that you are using? Is it the released 2.7, or current development code from GitHub? With the current code, USE_SERIAL_CONSOLE='no' should not be needed in most cases (after #2961 was merged).

**Comment by ufreier** [comment]
I use ReaR from the Debian repository (simply 'apt install rear') but I'll check the latest version from here.

**Comment by pcahyna** [comment]
@ufreier I believe this may explain it, please try the latest sources without USE_SERIAL_CONSOLE='no' ( here's how to use ReaR from Git sources instead from a package: https://github.com/rear/rear/#quick-start-guide )

**Comment by kronenpj** [comment]
I believe I was having the same problem and the code in git as of now (rear-2.7-1.git.2.0fd8a77a.master) works far better than 2.6 on Fedora 39. I look forward to this being picked up by Fedora's repositories. :)

**Comment by jsmeix** [comment]
Excerpts from
https://lists.freedesktop.org/archives/systemd-devel/2023-November/049676.html
```
Those libraries are not found by ldd because
they are not in the public library search path.
...
Those libraries have an unstable ABI,
so they are tucked away in a separate directory
and are not intended to be found during a search
```

When intentionally libraries cannot be found by 'ldd'
then such libraries must be explicitly specified
via LIBS, see its description in default.conf
e.g. for ReaR 2.7 online starting at
https://github.com/rear/rear/blob/rear-2.7/usr/share/rear/conf/default.conf#L1517

**Comment by tbsky** [comment]
Hi:
    I don't know if systemd is meant to be "Special library files" as described in the default.conf.  since systemd is used by most distributions, every user will need to handle that. I just thought if distribution packager or rear itself can do something so users won't need to take care of systemd issues in the future.

**Comment by jsmeix** [comment]
I meant that ReaR will have to be enhanced
to somehow deal with newer systemd's "hidden" libraries.
We have some code to include systemd into the ReaR recovery system
and that code would have to be somehow enhanced because that code
is meant to get some consistent systemd things (minimal as needed)  
into the ReaR recovery system.
Currently with newer systemd things are inconsistent in the
ReaR recovery system (some "hidden" libraries are missing)
so the missing things must be added to the ReaR recovery system.

Until the ReaR upstream GitHub master code was enhanced
to properly deal with newer systemd's "hidden" libraries
all users with newer systemd and older or current ReaR
need to manually specify the missing libraries via LIBS
as I described above in
https://github.com/rear/rear/issues/3021#issuecomment-1609096312

FWIW:
I think LD_PRELOAD stuff doesn't solve it because
I don't see how that makes the missing libraries
get included in the ReaR recovery system.
I fear LD_PRELOAD stuff only makes "rear recover" don't fail
but I don't see how that also makes "rear recover" work?
I don't see "rear recover" mentioned in the comments above.
But this is only what I think off the top of my head.
I did not test with newer systemd.
I only did
https://github.com/rear/rear/issues/3021#issuecomment-1609283508

**Comment by zDEFz** [comment]
> I meant that ReaR will have to be enhanced to somehow deal with newer systemd's "hidden" libraries. We have some code to include systemd into the ReaR recovery system and that code would have to be somehow enhanced because that code is meant to get some consistent systemd things (minimal as needed) into the ReaR recovery system. Currently with newer systemd things are inconsistent in the ReaR recovery system (some "hidden" libraries are missing) so the missing things must be added to the ReaR recovery system.
> 
> Until the ReaR upstream GitHub master code was enhanced to properly deal with newer systemd's "hidden" libraries all users with newer systemd and older or current ReaR need to manually specify the missing libraries via LIBS as I described above in [#3021 (comment)](https://github.com/rear/rear/issues/3021#issuecomment-1609096312)
> 
> FWIW: I think LD_PRELOAD stuff doesn't solve it because I don't see how that makes the missing libraries get included in the ReaR recovery system. I fear LD_PRELOAD stuff only makes "rear recover" don't fail but I don't see how that also makes "rear recover" work? I don't see "rear recover" mentioned in the comments above. But this is only what I think off the top of my head. I did not test with newer systemd. I only did [#3021 (comment)](https://github.com/rear/rear/issues/3021#issuecomment-1609283508) Even if LD_PRELOAD stuff solves it I think specifying the missing libraries via LIBS is the right method in ReaR that is meant for such things while LD_PRELOAD stuff would implement RFC1925 item (6a).


As said, the only thing fixing it for me was to edit the rear script and add 
` export LD_PRELOAD=/usr/lib64/systemd/libsystemd-shared-253.10-1.fc38.so `

**Comment by rasa** [comment]
To be clear, the command
```
sudo tee <<<'/usr/lib/x86_64-linux-gnu/systemd' /etc/ld.so.conf.d/x86_64-linux-gnu-systemd.conf
sudo ldconfig
```
removed the error message from the logs, on Ubuntu 23.10 (Mantic).

**Comment by schlomo** [comment]
It seems like a possible "root cause" could be the fact that our way of checking for hidden library dependencies with `cd` and `ldd` doesn't work in all cases
https://github.com/rear/rear/blob/58d6ec23941557af097def87ea25b13d340f834a/usr/share/rear/build/default/990_verify_rootfs.sh#L179

Maybe a solution would be to search for all `lib*` directories in the rescue system and use them as `LD_LIBRARY_PATH`? Or is that too extreme?

I think that there is value to removing such false warnings as they tend to confuse or scare users. And they make it so much harder to pick up real errors.

**Comment by schlomo** [comment]
@rear/contributors I think that this problem is somewhat specific to Systemd after a certain version, at least 252 and later show this. The general problem is "hidden" libraries and we won't be able to solve it in a generic fashion.

For quite some proprietary backup software we already added support for a "custom" `LD_LIBRARY_PATH` and I would suggest extending that mechanism to also include a customer `LD_LIBRARY_PATH` for Systemd to satisfy our `ldd` validation. It won't affect the rescue system if we do it only in ReaR itself for the validation of libraries.

Maybe in the future we will meet more "special" standard software that needs a custom library path, and then we should again accomodate it.

**Comment by GreasyMonkee** [comment]
> I found [this post](https://answers.launchpad.net/ubuntu/+question/703825) that offered some "solutions." Even though the question was asked for Ubuntu, I found it helpful for Debian.
> 
> TL;DR
> 
> One-off solution:
> 
> `sudo bash -c "LD_PRELOAD=/usr/lib/x86_64-linux-gnu/systemd/libsystemd-shared-252.so" rear -v mkbackup` (untested)
> 
> ... OR ...
> 
> create an file in `/etc/ld.so.conf.d` (This is the method I chose.)
> 
> I created `/etc/ld.so.conf.d/libsystemd-core.conf` ... into which I put `/usr/lib/x86_64-linux-gnu/systemd` I then ran `sudo ldconfig`
> 
> A run of `ldd /usr/lib/x86_64-linux-gnu/systemd/libsystemd-core-252.so` now resolves `libsystemd-shared-252.so`
> 
> I rebooted for good measure and ran a straight `rear -v mkbackup` -- it ran flawlessly.
> 
> I suppose why `libsystemd-core-252.so` has this issue is a question for the distro and subsystem teams.


Fedora 41 (Rawhide) using libsystemd-256-rc2-1 has the same behaviour.

Applied Liam Fry's fix of creating `/etc/ld.so.conf.d/libsystemd-core.conf` ... into which I put `/usr/lib64/systemd` I then ran `sudo ldconfig` has given a clean creation of the Backup image.

Many thanks for the whole team here for their contributions.

Cheers,
Garth.

**Comment by schlomo** [comment]
Thank you @GreasyMonkee for the test on Rawhide. I think that we should note that the `libsystemd-shared` library is *private* to systemd and therefore not part of the regular library search path. ReaR shows an error because ReaR checks all binaries for library dependencies and we need to extend that mechanism to also properly handle private libraries.

As a workaround I think it is OK to modify/hack ReaR - I'd caution against modifying the system in a way that works against the systemd design of keeping this library private. Maybe it has no negative side effect, but maybe it can do more harm somewhere else where you didn't expect it.

**Comment by GreasyMonkee** [comment]
> Thank you @GreasyMonkee for the test on Rawhide. I think that we should note that the `libsystemd-shared` library is _private_ to systemd and therefore not part of the regular library search path. ReaR shows an error because ReaR checks all binaries for library dependencies and we need to extend that mechanism to also properly handle private libraries.
> 
> As a workaround I think it is OK to modify/hack ReaR - I'd caution against modifying the system in a way that works against the systemd design of keeping this library private. Maybe it has no negative side effect, but maybe it can do more harm somewhere else where you didn't expect it.

Yes, agreed. I will be keeping an eye on how things behave over the coming months, if any issues crop up, shall let you know.

**Comment by jsmeix** [comment]
By chance I needed to try out ReaR on a Fedora 40 system, see
https://github.com/rear/rear/pull/3175#issuecomment-2163116469
and subsequent comments therein and see also
https://github.com/rear/rear/issues/3247

So I experienced this issue here myself.

What I found out:

With some recent systemd version
'rear -D mkrescue' shows something like:
```
Testing each binary with 'ldd' and look for 'not found' libraries within the recovery system
/usr/lib64/systemd/libsystemd-core-255.4-1.fc40.so requires additional libraries
    libsystemd-shared-255.4-1.fc40.so => not found
ReaR recovery system in '/var/tmp/rear.nHDN0gSTgC8zE80/rootfs' needs additional libraries, check /root/rear.github.master/var/log/rear/rear-linux.log for details
```

But
```
COPY_AS_IS+=( /usr/lib64/systemd/ )
```
does not help because those libraries
get already automatically included via
('rear -D mkrescue' log file excerpts)
```
+ source /root/rear.github.master/usr/share/rear/build/GNU/Linux/100_copy_as_is.sh
...
... Adding required library '/usr/lib64/systemd/libsystemd-core-255.4-1.fc40.so' to LIBS
... Adding required library '/usr/lib64/systemd/libsystemd-shared-255.4-1.fc40.so' to LIBS
```
So one gets them in the ReaR recovery system
```
# find /var/tmp/rear.nHDN0gSTgC8zE80/rootfs/ -type f | grep libsystemd-

/var/tmp/rear.nHDN0gSTgC8zE80/rootfs/usr/lib64/systemd/libsystemd-core-255.4-1.fc40.so
/var/tmp/rear.nHDN0gSTgC8zE80/rootfs/usr/lib64/systemd/libsystemd-shared-255.4-1.fc40.so
```

Also something like
```
NON_FATAL_BINARIES_WITH_MISSING_LIBRARY='systemd/libsystemd-core'
```
does not help to silence the message because
usr/share/rear/build/default/990_verify_rootfs.sh
has a comment that explains
```
# Only for programs (i.e. files in a .../bin/... or .../sbin/... directory) treat a missing library as fatal
```

So - at least for now - a message like
```
/usr/lib64/systemd/libsystemd-core-255.4-1.fc40.so requires additional libraries
    libsystemd-shared-255.4-1.fc40.so => not found
```
can only be ignored.

It is some kind of "false positive" but currently
I have no good idea how to avoid that without
breaking the intent of that verification step
i.e. without the risk of getting "false negatives"
which would be worse that this "false positive".

**Comment by jsmeix** [comment]
Only a totally untested offhanded idea
how to possibly solve such issues generically:

The crucial part here is that
libsystemd-shared is needed by libsystemd-core
but libsystemd-shared cannot be found as usual
regardless that libsystemd-shared exists in the
ReaR recovery system.

So my basic idea is to add a verification step
that tests whether or not a library that is needed
but cannot be found by usual means exists nevertheless
exists in the ReaR recovery system.

When it exists nevertheless exists in the recovery system
we may sufficiently safely consider things are OK.

**Comment by jsmeix** [comment]
With
https://github.com/rear/rear/pull/3250
merged this issue should be sufficiently solved.

**Comment by jsmeix** [comment]
A subsequent issue is
https://github.com/rear/rear/issues/3414
"Fix when 'find' results only dangling symlinks to needed libraries"
cf.
https://github.com/rear/rear/pull/3308#issuecomment-2328139914
```
... it is not sufficient when something
that is reported as 'not found' by 'ldd'
actually exists within the recovery system
to assume that then the associated program(s)
will actually work within the recovery system ...
```
